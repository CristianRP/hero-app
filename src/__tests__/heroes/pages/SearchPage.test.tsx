import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../heroes';

const mockUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockUseNavigate
}));

describe('Tests on <SearcPage /> component', () => {

  beforeEach(() => jest.clearAllMocks() );

  test('should show default values', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    )

    expect( container ).toMatchSnapshot();
  });

  test('should show batman and input with queryString', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>
    )

    const input = screen.getByRole<HTMLInputElement>('textbox');
    expect( input.value ).toBe('batman');

    const img = screen.getByRole<HTMLImageElement>('img');
    expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
  });

  test('should show an error if the hero is not found', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=abc']}>
        <SearchPage />
      </MemoryRouter>
    );

    expect( screen.getByRole('alert') ).toBeTruthy();
    expect( screen.getByRole('alert').innerHTML ).toBe('No hero with <b>abc</b>');
  });

  test('should navigate to the new page', () => {
    const hero = 'flash';

    render(
      <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
        <SearchPage />
      </MemoryRouter> 
    );

    fireEvent.click( screen.getByRole('button', { name: 'Search' }) );

    expect( mockUseNavigate ).toHaveBeenCalled();
    expect( mockUseNavigate ).toHaveBeenCalledWith(`?q=${hero}`);
  });
});