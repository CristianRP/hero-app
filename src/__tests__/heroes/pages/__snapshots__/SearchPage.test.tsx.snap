// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearcPage /> component should show default values 1`] = `
<div>
  <h1
    class="text-3xl font-bold mt-5"
  >
    Search
  </h1>
  <hr
    class="mt-5 mb-5"
  />
  <div
    class="flex flex-wrap mt-5"
  >
    <div
      class="w-full md:w-4/12"
    >
      <h4
        class="text-xl font-semibold"
      >
        Searching
      </h4>
      <hr
        class="my-2"
      />
      <form
        class="flex flex-col"
      >
        <input
          autocomplete="off"
          class="flex-1 rounded-sm"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="w-1/3 mt-2 px-4 py-2 bg-blue-500 text-white rounded-md hover:bg-blue-600 focus:outline-none focus:bg-blue-600 focus:ring focus:ring-blue-300"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="w-full md:w-4/6 px-4"
    >
      <h4
        class="text-xl font-semibold"
      >
        Results
      </h4>
      <hr
        class="my-2"
      />
      <div
        class="bg-blue-100 rounded-md border border-blue-500 text-blue-700 p-4 mb-4 animate__animated animate__fadeIn"
        role="alert"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
